File: native-ios/Solvio/Core/ToastCenter.swift
import Foundation
import SwiftUI

/// Lightweight top-banner toast surface. Matches the PWA's `sonner`
/// toaster — success/info/error/warning with icon + message.
@MainActor
final class ToastCenter: ObservableObject {
    enum Kind {
        case success
        case error
        case info
        case warning
    }

    /// Optional undo action attached to a toast. When present the toast
    /// renders an "Undo" button — tapping it cancels the auto-dismiss timer
    /// and runs `handler`. Use cases: bulk delete, swipe-to-delete, archive.
    struct UndoAction: Equatable {
        let label: String
        let handler: () -> Void
        // Equatable conformance: two undo actions are "equal" when they
        // share the same label — handlers are closures and can't be
        // compared. Good enough for SwiftUI animation diffing.
        static func == (lhs: UndoAction, rhs: UndoAction) -> Bool { lhs.label == rhs.label }
    }

    struct Toast: Identifiable, Equatable {
        let id = UUID()
        let kind: Kind
        let title: String
        let description: String?
        var undo: UndoAction? = nil
    }

    @Published var current: Toast?

    func success(_ title: String, description: String? = nil) {
        show(Toast(kind: .success, title: title, description: description))
    }

    func error(_ title: String, description: String? = nil) {
        show(Toast(kind: .error, title: title, description: description))
    }

    func info(_ title: String, description: String? = nil) {
        show(Toast(kind: .info, title: title, description: description))
    }

    func warning(_ title: String, description: String? = nil) {
        show(Toast(kind: .warning, title: title, description: description))
    }

    /// Show an info toast with an undo affordance. Kept open ~5 s so the
    /// user has time to react before the destructive action commits.
    func undoable(_ title: String, undoLabel: String, undo: @escaping () -> Void) {
        show(Toast(kind: .info, title: title, description: nil, undo: UndoAction(label: undoLabel, handler: undo)))
    }

    private func show(_ toast: Toast) {
        current = toast
        // Undo toasts get the longer 5 s window so the user has time to act.
        let isUndoOrError = toast.kind == .error || toast.undo != nil
        let displayNs: UInt64 = isUndoOrError ? 5_000_000_000 : 3_200_000_000
        Task {
            try? await Task.sleep(nanoseconds: displayNs)
            if current?.id == toast.id {
                current = nil
            }
        }
    }

    func dismiss() { current = nil }

    /// Run the toast's undo handler if present and dismiss. Called from the
    /// toast UI when the user taps the inline "Undo" button.
    func performUndo() {
        guard let undo = current?.undo else { return }
        undo.handler()
        dismiss()
    }
}

struct ToastOverlay: View {
    @EnvironmentObject private var toast: ToastCenter

    var body: some View {
        VStack {
            if let t = toast.current {
                HStack(alignment: .top, spacing: Theme.Spacing.sm) {
                    icon(for: t.kind)
                        .frame(width: 20, height: 20)
                    VStack(alignment: .leading, spacing: 2) {
                        Text(t.title)
                            .font(AppFont.bodyMedium)
                            .foregroundColor(Theme.foreground)
                        if let d = t.description {
                            Text(d)
                                .font(AppFont.caption)
                                .foregroundColor(Theme.mutedForeground)
                                .fixedSize(horizontal: false, vertical: true)
                        }
                    }
                    Spacer()
                    // Undo button — explicit Button (not onTapGesture) so it
                    // captures the tap before the toast-level dismiss runs.
                    if let u = t.undo {
                        Button(action: { toast.performUndo() }) {
                            Text(u.label)
                                .font(AppFont.mono(11))
                                .tracking(1)
                                .textCase(.uppercase)
                                .padding(.horizontal, 10)
                                .padding(.vertical, 6)
                                .foregroundColor(Theme.background)
                                .background(Theme.foreground)
                                .clipShape(RoundedRectangle(cornerRadius: Theme.Radius.sm))
                        }
                        .buttonStyle(.plain)
                    }
                }
                .padding(Theme.Spacing.md)
                .nbCard(radius: Theme.Radius.md, shadow: Theme.Shadow.md)
                .padding(.horizontal, Theme.Spacing.md)
                .transition(.move(edge: .top).combined(with: .opacity))
                // Tap-to-dismiss only on toasts WITHOUT an undo affordance.
                // For undoable toasts a stray tap on the body would dismiss
                // and silently lose the undo opportunity — surprise loss.
                .onTapGesture {
                    if t.undo == nil { toast.dismiss() }
                }
            }
            Spacer()
        }
        .padding(.top, Theme.Spacing.sm)
        .animation(.spring(response: 0.35, dampingFraction: 0.85), value: toast.current)
        .allowsHitTesting(toast.current != nil)
    }

    @ViewBuilder
    private func icon(for kind: ToastCenter.Kind) -> some View {
        switch kind {
        case .success:
            Image(systemName: "checkmark.circle.fill").foregroundColor(Theme.success)
        case .error:
            Image(systemName: "exclamationmark.octagon.fill").foregroundColor(Theme.destructive)
        case .warning:
            Image(systemName: "exclamationmark.triangle.fill").foregroundColor(Theme.warning)
        case .info:
            Image(systemName: "info.circle.fill").foregroundColor(Theme.info)
        }
    }
}

File: native-android/app/src/main/java/com/programo/solvio/core/toast/ToastCenter.kt
package com.programo.solvio.core.toast

import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.util.UUID

/** Mirror of iOS [ToastCenter]. */
class ToastCenter {
    enum class Kind { Success, Error, Info, Warning }

    data class UndoAction(val label: String, val handler: () -> Unit)

    data class Toast(
        val id: String = UUID.randomUUID().toString(),
        val kind: Kind,
        val title: String,
        val description: String? = null,
        val undo: UndoAction? = null,
    )

    private val scope: CoroutineScope = MainScope()
    var current by mutableStateOf<Toast?>(null)
        private set

    fun success(title: String, description: String? = null) =
        show(Toast(kind = Kind.Success, title = title, description = description))

    fun error(title: String, description: String? = null) =
        show(Toast(kind = Kind.Error, title = title, description = description))

    fun info(title: String, description: String? = null) =
        show(Toast(kind = Kind.Info, title = title, description = description))

    fun warning(title: String, description: String? = null) =
        show(Toast(kind = Kind.Warning, title = title, description = description))

    fun undoable(title: String, undoLabel: String, undo: () -> Unit) =
        show(Toast(kind = Kind.Info, title = title, undo = UndoAction(undoLabel, undo)))

    fun dismiss() { current = null }

    fun performUndo() {
        val u = current?.undo ?: return
        u.handler()
        dismiss()
    }

    private fun show(toast: Toast) {
        current = toast
        val displayMs = if (toast.kind == Kind.Error || toast.undo != null) 5_000L else 3_200L
        scope.launch {
            delay(displayMs)
            if (current?.id == toast.id) current = null
        }
    }
}

val LocalToast = compositionLocalOf<ToastCenter> { error("ToastCenter not provided") }
